allow op-node to retry fetching data at latest height with light node

<img width="3472" alt="scrot 2023-05-02 at 9 43 54 AM" src="https://user-images.githubusercontent.com/46639943/235684684-083c3263-dbf8-4cda-963a-b2b9fded5dea.png">

```bash
t=2023-05-02T13:43:40+0000 lvl=warn msg="Served optimism_syncStatus"             conn=172.20.0.7:53456 reqid=553 duration=1.999226376s err="context canceled"
t=2023-05-02T13:43:41+0000 lvl=warn msg="Served optimism_syncStatus"             conn=172.20.0.6:37246 reqid=501 duration=2.000833792s err="context canceled"
t=2023-05-02T13:43:42+0000 lvl=warn msg="Served optimism_syncStatus"             conn=172.20.0.7:53460 reqid=554 duration=1.999152876s err="context canceled"
t=2023-05-02T13:43:43+0000 lvl=warn msg="Served optimism_syncStatus"             conn=172.20.0.6:37252 reqid=502 duration=1.999913626s err="context canceled"
t=2023-05-02T13:43:44+0000 lvl=warn msg="Served optimism_syncStatus"             conn=172.20.0.7:53476 reqid=555 duration=2.00052196s  err="context canceled"
t=2023-05-02T13:43:45+0000 lvl=warn msg="unable to retrieve data from da"        origin=0x16c579f9a0236163eaeb581267fe75e7ac3379513adcf2b6c4655491f97046b5:392 err="Get \"http://da:26659/namespaced_data/e8e5f679bf7116cb/height/0\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
panic: runtime error: index out of range [0] with length 0

goroutine 238 [running]:
github.com/ethereum-optimism/optimism/op-node/rollup/derive.DataFromEVMTransactions(0x400018a7d0, 0x400018a900, {0x3c, 0x44, 0xcd, 0xdd, 0xb6, 0xa9, 0x0, 0xfa, ...}, ...)
	/app/op-node/rollup/derive/calldata_source.go:136 +0x6f8
github.com/ethereum-optimism/optimism/op-node/rollup/derive.NewDataSource({0x1c9fb28?, 0x4000042030?}, {0x1cadfa0?, 0x400034c940}, 0x400018a7d0, 0xc94fac?, {0xffff7c8a2de8, 0x4000bf09c0}, {{0x16, 0xc5, ...}, ...}, ...)
	/app/op-node/rollup/derive/calldata_source.go:80 +0x260
github.com/ethereum-optimism/optimism/op-node/rollup/derive.(*DataSourceFactory).OpenData(0x0?, {0x1c9fb28?, 0x4000042030?}, {{0x16, 0xc5, 0x79, 0xf9, 0xa0, 0x23, 0x61, ...}, ...}, ...)
	/app/op-node/rollup/derive/calldata_source.go:44 +0x88
github.com/ethereum-optimism/optimism/op-node/rollup/derive.(*L1Retrieval).NextData(0x4001010c80, {0x1c9fb28, 0x4000042030})
	/app/op-node/rollup/derive/l1_retrieval.go:56 +0x194
github.com/ethereum-optimism/optimism/op-node/rollup/derive.(*FrameQueue).NextFrame(0x4001010cc0, {0x1c9fb28?, 0x4000042030?})
	/app/op-node/rollup/derive/frame_queue.go:39 +0x60
github.com/ethereum-optimism/optimism/op-node/rollup/derive.(*ChannelBank).NextData(0x400100db60, {0x1c9fb28, 0x4000042030})
	/app/op-node/rollup/derive/channel_bank.go:154 +0xe8
github.com/ethereum-optimism/optimism/op-node/rollup/derive.(*ChannelInReader).NextBatch(0x4002088780, {0x1c9fb28?, 0x4000042030?})
	/app/op-node/rollup/derive/channel_in_reader.go:66 +0x40
github.com/ethereum-optimism/optimism/op-node/rollup/derive.(*BatchQueue).NextBatch(0x40000a8780, {0x1c9fb28, 0x4000042030}, {{0x65, 0x58, 0xf7, 0x59, 0xda, 0x46, 0xf6, ...}, ...})
	/app/op-node/rollup/derive/batch_queue.go:87 +0x354
github.com/ethereum-optimism/optimism/op-node/rollup/derive.(*AttributesQueue).NextAttributes(0x4001010d00, {0x1c9fb28, 0x4000042030}, {{0x65, 0x58, 0xf7, 0x59, 0xda, 0x46, 0xf6, ...}, ...})
	/app/op-node/rollup/derive/attributes_queue.go:54 +0x68
github.com/ethereum-optimism/optimism/op-node/rollup/derive.(*EngineQueue).Step(0x4000780800, {0x1c9fb28, 0x4000042030})
	/app/op-node/rollup/derive/engine_queue.go:243 +0x134
github.com/ethereum-optimism/optimism/op-node/rollup/derive.(*DerivationPipeline).Step(0x400016e930, {0x1c9fb28, 0x4000042030})
	/app/op-node/rollup/derive/pipeline.go:199 +0x4a0
github.com/ethereum-optimism/optimism/op-node/rollup/driver.(*Driver).eventLoop(0x40020b4000)
	/app/op-node/rollup/driver/state.go:291 +0x1120
created by github.com/ethereum-optimism/optimism/op-node/rollup/driver.(*Driver).Start
	/app/op-node/rollup/driver/state.go:92 +0x84
```

--- a/op-node/rollup/derive/calldata_source.go
+++ b/op-node/rollup/derive/calldata_source.go
@@ -75,9 +75,20 @@ func NewDataSource(ctx context.Context, log log.Logger, cfg *rollup.Config, daCf
 			batcherAddr: batcherAddr,
 		}
 	} else {
+		data, err := DataFromEVMTransactions(cfg, daCfg, batcherAddr, txs, log.New("origin", block))
+		if err != nil {
+			return &DataSource{
+				open:        false,
+				id:          block,
+				cfg:         cfg,
+				fetcher:     fetcher,
+				log:         log,
+				batcherAddr: batcherAddr,
+			}
+		}
 		return &DataSource{
 			open: true,
-			data: DataFromEVMTransactions(cfg, daCfg, batcherAddr, txs, log.New("origin", block)),
+			data: data,
 		}
 	}
 }
@@ -89,7 +100,10 @@ func (ds *DataSource) Next(ctx context.Context) (eth.Data, error) {
 	if !ds.open {
 		if _, txs, err := ds.fetcher.InfoAndTxsByHash(ctx, ds.id.Hash); err == nil {
 			ds.open = true
-			ds.data = DataFromEVMTransactions(ds.cfg, ds.daCfg, ds.batcherAddr, txs, log.New("origin", ds.id))
+			ds.data, err = DataFromEVMTransactions(ds.cfg, ds.daCfg, ds.batcherAddr, txs, log.New("origin", ds.id))
+			if err != nil {
+				return nil, NewTemporaryError(fmt.Errorf("failed to open calldata source: %w", err))
+			}
 		} else if errors.Is(err, ethereum.NotFound) {
 			return nil, NewResetError(fmt.Errorf("failed to open calldata source: %w", err))
 		} else {
@@ -108,7 +122,7 @@ func (ds *DataSource) Next(ctx context.Context) (eth.Data, error) {
 // DataFromEVMTransactions filters all of the transactions and returns the calldata from transactions
 // that are sent to the batch inbox address from the batch sender address.
 // This will return an empty array if no valid transactions are found.
-func DataFromEVMTransactions(config *rollup.Config, daCfg *rollup.DAConfig, batcherAddr common.Address, txs types.Transactions, log log.Logger) []eth.Data {
+func DataFromEVMTransactions(config *rollup.Config, daCfg *rollup.DAConfig, batcherAddr common.Address, txs types.Transactions, log log.Logger) ([]eth.Data, error) {
 	var out []eth.Data
 	l1Signer := config.L1Signer()
 	for j, tx := range txs {
@@ -124,19 +138,24 @@ func DataFromEVMTransactions(config *rollup.Config, daCfg *rollup.DAConfig, batc
 				continue // not an authorized batch submitter, ignore
 			}
 
-			height, index, err := decodeETHData(tx.Data())
-			if err != nil {
-				log.Warn("unable to decode data pointer", "index", j, "err", err)
-				continue
+			if daCfg != nil {
+				height, index, err := decodeETHData(tx.Data())
+				if err != nil {
+					log.Warn("unable to decode data pointer", "index", j, "err", err)
+					continue
+				}
+				data, err := daCfg.Client.NamespacedData(context.Background(), daCfg.NamespaceId, uint64(height))
+				if err != nil {
+					log.Warn("unable to retrieve data from da", "err", err)
+					return nil, err
+				}
+				out = append(out, data[index])
+			} else {
+				out = append(out, tx.Data())
 			}
-			data, err := daCfg.Client.NamespacedData(context.Background(), daCfg.NamespaceId, uint64(height))
-			if err != nil {
-				log.Warn("unable to retrieve data from da", "err", err)
-			}
-			out = append(out, data[index])
 		}
 	}
-	return out
+	return out, nil
 }
 
 // decodeETHData will decode the data retrieved from the EVM, this data

--- a/op-node/rollup/derive/calldata_source_test.go
+++ b/op-node/rollup/derive/calldata_source_test.go
@@ -121,8 +121,9 @@ func TestDataFromEVMTransactions(t *testing.T) {
 			}
 		}
 
-		out := DataFromEVMTransactions(cfg, batcherAddr, txs, testlog.Logger(t, log.LvlCrit))
+		out, err := DataFromEVMTransactions(cfg, nil, batcherAddr, txs, testlog.Logger(t, log.LvlCrit))
 		require.ElementsMatch(t, expectedData, out)
+		require.NoError(t, err)
 	}
 
 }

